[LLPC] Scalarize non-uniform loads inside the waterfall loop

kmitropoulou opened the pull request.

jayfoad commented:
Can you add a description please? Is this PR identical to #2645 or have you fixed something?

kmitropoulou commented:
Yes, this patch is an update of #2645

amdvlk-admin commented:
### Test summary for commit 7033caf0ad3b496eaf5ade9ff31fc067bc9033e6
<details><summary>CTS tests (Failed: 31/139021)</summary><ul><li>Built with version 1.3.5.2</li>
<details open><summary>Ubuntu navi3x, Srdcvk</summary><ul>

<li>Passed: 41128/69517 (59.2%)</li>
<li><details><summary>Failed: 16/69517 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.descriptor_indexing.sampled_image
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_after_bind
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_after_bind_in_loop
Stack trace: Flake

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_after_bind_in_loop_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_after_bind_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_in_loop
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_in_loop_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_after_bind
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_after_bind_in_loop
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_after_bind_in_loop_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_after_bind_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_in_loop
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_in_loop_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_with_lod
Stack trace: Fail
```
</p></details></li>

<li>Not Supported: 28373/69517 (40.8%)</li><li>Warnings: 0/69517 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu navi2x, Srdcvk</summary><ul>

<li>Passed: 41163/69504 (59.2%)</li>
<li><details><summary>Failed: 15/69504 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.descriptor_indexing.sampled_image
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_after_bind
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_after_bind_in_loop_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_after_bind_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_in_loop
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_in_loop_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_after_bind
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_after_bind_in_loop
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_after_bind_in_loop_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_after_bind_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_in_loop
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_in_loop_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_with_lod
Stack trace: Fail
```
</p></details></li>

<li>Not Supported: 28326/69504 (40.8%)</li><li>Warnings: 0/69504 (0.0%)</li>
</ul></details>


</ul></details>

kmitropoulou commented:
test this please

perlfu commented:
Technically this is a revert for #2705 as well?
As it brings back the optional_bool parts, etc.

At a high level, GFX 10.3.2 codegen is just modifying the image descriptor after it is loaded as a workaround.
This code seems to explicitly match the sequence of that modification?
Is it possible to make the code more generic so it would be robust to changes in that modification sequence?

kmitropoulou commented:
>> Technically this is a revert for https://github.com/GPUOpen-Drivers/llpc/pull/2705 as well?

#2645 should have never been committed. So, I had to revert it in #2705.

amdvlk-admin commented:
### Test summary for commit 7033caf0ad3b496eaf5ade9ff31fc067bc9033e6
<details><summary>CTS tests (Failed: 31/139021)</summary><ul><li>Built with version 1.3.5.2</li>
<details open><summary>Ubuntu navi3x, Srdcvk</summary><ul>

<li>Passed: 41129/69517 (59.2%)</li>
<li><details><summary>Failed: 15/69517 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.descriptor_indexing.sampled_image
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_after_bind
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_after_bind_in_loop_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_after_bind_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_in_loop
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_in_loop_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_after_bind
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_after_bind_in_loop
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_after_bind_in_loop_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_after_bind_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_in_loop
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_in_loop_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_with_lod
Stack trace: Fail
```
</p></details></li>

<li>Not Supported: 28373/69517 (40.8%)</li><li>Warnings: 0/69517 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu navi2x, Srdcvk</summary><ul>

<li>Passed: 41162/69504 (59.2%)</li>
<li><details><summary>Failed: 16/69504 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.descriptor_indexing.sampled_image
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_after_bind
Stack trace: Flake

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_after_bind_in_loop
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_after_bind_in_loop_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_after_bind_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_in_loop
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_in_loop_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_after_bind
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_after_bind_in_loop
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_after_bind_in_loop_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_after_bind_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_in_loop
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_in_loop_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_with_lod
Stack trace: Fail
```
</p></details></li>

<li>Not Supported: 28326/69504 (40.8%)</li><li>Warnings: 0/69504 (0.0%)</li>
</ul></details>


</ul></details>

nhaehnle commented on lgc/builder/BuilderImpl.cpp:
```diff
@@ -503,36 +505,115 @@ static Value *traceNonUniformIndex(Value *nonUniformVal) {
+    Value *OrigLoad = nullptr;
+    Value *OrigAnd = nullptr;
+    Value *OrigGep = nullptr;
+    Value *OrigExtract = nullptr;
```
Coding style in LLPC is to use lowerCamelCase.

kmitropoulou replied:
Done :)

nhaehnle reviewed (commented):
The use of matching feels sort of backwards. If we know the code sequences that should go into the waterfall loops, shouldn't we create them like that directly? I'll try to see if I can expand on this, but I think the gist of it is that we should put the waterfall loop creation upside-down: instead of creating the image instruction and then calling createWaterfallLoop, we should call createWaterfallLoop *first*, and then insert instructions into the loop.

nhaehnle reviewed (commented):
I remember now what the difficulty was with just creating waterfall loops directly. The descriptors load instructions are created separately from when the actual image access instructions are created. I wonder whether we shouldn't change this.

Meanwhile, this PR is rather fragile and really shouldn't be adopted as-is. Moving instructions around without checking all their uses is a recipe for creating invalid IR.

nhaehnle commented on lgc/include/lgc/builder/BuilderImpl.h:
```diff
@@ -58,6 +58,16 @@ class BuilderImpl : public BuilderDefs {
                                          bool scalarizeDescriptorLoads = false, bool useVgprForOperands = false,
                                          const llvm::Twine &instName = "");
 
+  llvm::Value *emitWaterfallBegin(llvm::Instruction *nonUniformInst, llvm::ArrayRef<unsigned> operandIdxs,
+                                  llvm::ArrayRef<llvm::Value *> nonUniformIndices, bool useVgprForOperands = false,
+                                  const llvm::Twine &instName = "");
+
+  llvm::Value *emitWaterfallBeginForScalarizedLoops(llvm::Instruction *nonUniformInst,
```
These two methods should be private.

kmitropoulou replied:
Done :)

nhaehnle commented on lgc/builder/BuilderImpl.cpp:
```diff
@@ -559,9 +640,13 @@ Instruction *BuilderImpl::createWaterfallLoop(Instruction *nonUniformInst, Array
     return nonUniformInst;
 
+  if (nonUniformInst->getType()->isVoidTy())
+    scalarizeDescriptorLoads = false;
```
Why this?

kmitropoulou replied:
I am sorry I missed this. This is for the case that we have the following intrinsic: @llvm.amdgcn.struct.buffer.store.format.v4f32()

tsymalla replied:
Can you please comment on that?

kmitropoulou replied:
Done

nhaehnle replied:
This still doesn't make sense though. The "scalarization" is a function of the *descriptor inputs* to the intrinsic, not a function of its output. This should be removed.

kmitropoulou replied:
I remove it and I added two tests : scalarizationOfDescriptorLoadTest1.ll and scalarizationOfDescriptorLoadTest2.ll

nhaehnle commented on lgc/builder/BuilderImpl.cpp:
```diff
@@ -503,36 +505,115 @@ static Value *traceNonUniformIndex(Value *nonUniformVal) {
+      newLoad->insertBefore(nonUniformInst);
+      nonUniformInst->setOperand(operandIdx, newLoad);
+      newGep->insertBefore(newLoad);
+      newGep->setOperand(1, readFirstLane);
```
Need to check that operand 1 is actually the firstIndexInst.

kmitropoulou replied:
I think we should not do it in the new version of the code.

nhaehnle commented on lgc/builder/BuilderImpl.cpp:
```diff
@@ -503,36 +505,115 @@ static Value *traceNonUniformIndex(Value *nonUniformVal) {
+      auto *loadOp = dyn_cast<LoadInst>(OrigLoad);
+      auto *gepOp = dyn_cast<GetElementPtrInst>(OrigGep);
+      auto *AndOp = dyn_cast<Instruction>(OrigAnd);
+      auto *extractOp = dyn_cast<ExtractElementInst>(OrigExtract);
```
These should all just be `cast`. No reason to use `dyn_cast` after previously doing all the match checks.

kmitropoulou replied:
Thank you for pointing out :) Not needed any more.

nhaehnle reviewed (commented):
I suppose with some fixes this PR can be at least made correct.

nhaehnle commented:
Yet another follow-up to this... wouldn't it be possible to collect the relevant instructions in `traceNonUniformIndex` and then optionally clone them into the waterfall loop in a generic fashion? That seems preferable over hard-coding the specific code sequences.

perlfu commented:
> Yet another follow-up to this... wouldn't it be possible to collect the relevant instructions in `traceNonUniformIndex` and then optionally clone them into the waterfall loop in a generic fashion? That seems preferable over hard-coding the specific code sequences.

I agree capturing the visited instruction chain from `traceNonUniformIndex` seems like a good way to avoid this being brittle.

Longer term I have been suggesting that we move this lowering to separate pass later in the compiler and simply annotate the output of non-uniform  instructions as `%val = call @lgc.create.waterfall(%ret)` or similar.
Your point about earlier parts of the compiler knowing potentially knowing which instructions are non-uniform is valid.
We could be preserving those annotations out of the SPIRVReader for use in later lowering -- although I suspect in some cases it might not tell us much more than `operandIdxs` does already.

kmitropoulou commented:
@nhaehnle : Thank you for the suggestion. I will try to capture the instruction chain from traceNonUniformIndex .

tsymalla-AMD commented on lgc/builder/BuilderImpl.cpp:
```diff
@@ -35,9 +35,11 @@
 #include "lgc/state/TargetInfo.h"
 #include "llvm/IR/IntrinsicInst.h"
 #include "llvm/IR/IntrinsicsAMDGPU.h"
+#include "llvm/IR/PatternMatch.h"
```
What is this being used for?

tsymalla-AMD replied:
Ah, I see. Can you use the namespace directly into the method `emitWaterfallBeginForScalarizedLoops`?

kmitropoulou commented:
@tsymalla-AMD : I removed all the pattern match code :)

tsymalla-AMD commented on lgc/builder/BuilderImpl.cpp:
```diff
@@ -336,49 +336,67 @@ BranchInst *BuilderImpl::createIf(Value *condition, bool wantElse, const Twine &
+// This uses a fairly simple heuristic that nevertheless allows temporary expansion of the search breadth to handle
+// the common case where a base pointer is assembled from separate high and low halves.
+//
+// @param nonUniformVal : Value representing non-uniform descriptor
```
Add `@param load`

kmitropoulou replied:
Done.

tsymalla-AMD commented on lgc/builder/BuilderImpl.cpp:
```diff
@@ -336,49 +336,67 @@ BranchInst *BuilderImpl::createIf(Value *condition, bool wantElse, const Twine &
+// the common case where a base pointer is assembled from separate high and low halves.
+//
+// @param nonUniformVal : Value representing non-uniform descriptor
+// @return : Value representing the non-uniform index, or null if nonUniformVal could be proven to be uniform
```
```suggestion
// @return : Value representing the non-uniform index, or nullptr if nonUniformVal could be proven to be uniform
```

kmitropoulou replied:
Done.

tsymalla-AMD commented on lgc/builder/BuilderImpl.cpp:
```diff
@@ -503,36 +521,130 @@ static Value *traceNonUniformIndex(Value *nonUniformVal) {
 
   return candidateIndex;
 }
+#endif
```
Add a comment on the condition of the #if

kmitropoulou replied:
Done.

tsymalla-AMD commented on lgc/builder/BuilderImpl.cpp:
```diff
@@ -503,36 +521,130 @@ static Value *traceNonUniformIndex(Value *nonUniformVal) {
-      return false;
-    if (!lhsInst->isIdenticalTo(rhsInst))
-      return false;
+Value *BuilderImpl::emitWaterfallBegin(Instruction *nonUniformInst, ArrayRef<unsigned> operandIdxs,
```
Can you add a comment to this method, describing its functionality and the arguments?

kmitropoulou replied:
Done.

tsymalla-AMD commented on lgc/builder/BuilderImpl.cpp:
```diff
@@ -503,36 +521,130 @@ static Value *traceNonUniformIndex(Value *nonUniformVal) {
+  Value *waterfallBegin = nullptr;
+
+  // Insert new code just before nonUniformInst.
+  SetInsertPoint(nonUniformInst);
```
What if this is called with nonUniformInst == nullptr?

kmitropoulou replied:
I am not sure if this is needed because the watefall loop is emitted for nonUniformInst.

tsymalla-AMD commented on lgc/builder/BuilderImpl.cpp:
```diff
@@ -503,36 +521,130 @@ static Value *traceNonUniformIndex(Value *nonUniformVal) {
+
+  // The first begin contains a null token for the previous token argument
+  waterfallBegin = ConstantInt::get(getInt32Ty(), 0);
+  for (auto nonUniformVal : nonUniformIndices)
```
Rather use `nonUniformIndex`

kmitropoulou replied:
Done.

tsymalla-AMD commented on lgc/builder/BuilderImpl.cpp:
```diff
@@ -503,36 +521,130 @@ static Value *traceNonUniformIndex(Value *nonUniformVal) {
+      return emitWaterfallBegin(nonUniformInst, operandIdxs, nonUniformIndices, useVgprForOperands, instName);
+  }
+
+  assert(firstIndexInst);
```
Assert for both `nonUniformInst` and `firstIndexInst`?

kmitropoulou replied:
I added it. I am not sure if it is needed.

tsymalla-AMD commented on lgc/builder/BuilderImpl.cpp:
```diff
@@ -503,36 +521,130 @@ static Value *traceNonUniformIndex(Value *nonUniformVal) {
+      newGep->setOperand(1, readFirstLane);
+      newLoad->setOperand(0, newGep);
+    } else if (isa<InsertElementInst>(desc)) {
+      // This case is for gfx 10.3.2
```
Shouldn't this be guarded then?

kmitropoulou replied:
Why do we need to guard it?

tsymalla-AMD commented on lgc/builder/BuilderImpl.cpp:
```diff
@@ -503,36 +521,130 @@ static Value *traceNonUniformIndex(Value *nonUniformVal) {
+      for (auto [origInst, newInst] : clonedOrigValuesMap) {
+        for (Use &use : newInst->operands()) {
+          Value *op = use.get();
+          if (!isa<Instruction>(op))
```
Can use `dyn_cast`

kmitropoulou replied:
Done.

tsymalla-AMD reviewed (changes requested):
Looks like some CTS are failing, but I did not check if those are related.

kmitropoulou commented on lgc/builder/BuilderImpl.cpp:
```diff
@@ -503,36 +521,130 @@ static Value *traceNonUniformIndex(Value *nonUniformVal) {
+          if (!isa<Instruction>(op))
+            continue;
+          auto *opI = cast<Instruction>(op);
+          auto It = clonedOrigValuesMap.find(opI);
```
Done.

kmitropoulou commented on lgc/builder/BuilderImpl.cpp:
```diff
@@ -554,14 +666,21 @@ Instruction *BuilderImpl::createWaterfallLoop(Instruction *nonUniformInst, Array
-    Value *nonUniformIndex = traceNonUniformIndex(nonUniformInst->getOperand(operandIdx));
+    Value *nonUniformVal = nonUniformInst->getOperand(operandIdx);
+    LoadInst *load = getLoad(nonUniformVal, loadChain);
+    Value *nonUniformIndex = !load ? nonUniformVal : traceNonUniformIndex(nonUniformVal, load);
```
Done.

amdvlk-admin commented:
### Test summary for commit d180bcf171ceb6d0d258fe681c46b05528a91c42
<details><summary>CTS tests (Failed: 32/138994)</summary><ul><li>Built with version 1.3.5.2</li>
<details open><summary>Ubuntu navi3x, Srdcvk</summary><ul>

<li>Passed: 41128/69503 (59.2%)</li>
<li><details><summary>Failed: 16/69503 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.descriptor_indexing.sampled_image
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_after_bind
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_after_bind_in_loop
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_after_bind_in_loop_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_after_bind_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_in_loop
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_in_loop_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_after_bind
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_after_bind_in_loop
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_after_bind_in_loop_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_after_bind_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_in_loop
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_in_loop_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_with_lod
Stack trace: Fail
```
</p></details></li>

<li>Not Supported: 28359/69503 (40.8%)</li><li>Warnings: 0/69503 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu navi2x, Srdcvk</summary><ul>

<li>Passed: 41162/69491 (59.2%)</li>
<li><details><summary>Failed: 16/69491 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.descriptor_indexing.sampled_image
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_after_bind
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_after_bind_in_loop
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_after_bind_in_loop_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_after_bind_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_in_loop
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_in_loop_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampled_image_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_after_bind
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_after_bind_in_loop
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_after_bind_in_loop_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_after_bind_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_in_loop
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_in_loop_with_lod
Stack trace: Fail

FAILURE: dEQP-VK.descriptor_indexing.sampler_with_lod
Stack trace: Fail
```
</p></details></li>

<li>Not Supported: 28313/69491 (40.7%)</li><li>Warnings: 0/69491 (0.0%)</li>
</ul></details>


</ul></details>

nhaehnle commented:
There are still some test failures. Can you please take a look at that?

nhaehnle reviewed (commented):
Thanks for looking into making it more generic / capturing the instructions only once. I haven't looked in detail yet, , but the fact that `emitWaterfallBeginForScalarizedLoops` still has some case distinctions is suspicious.

Keep in mind that you should think about the graph of data flow in clear terms. There is a non-uniform operand to the key instruction (e.g. image load/store). Call it X. We're interested in the subgraph of data flow that feeds into X. Basically, this whole "scalarization" thing is really about finding a good cut in the data flow graph -- the cut is where the readfirstlane is ultimately inserted:

* The naive cut is to just cut at X. This causes readfirstlane of the X value. This is correct, but the problem is that X may be big (a descriptor -- 8 dwords).
* We want to find a better cut, which primarily means finding a smaller cut. Ideally just 1 dword big.
* In order to do that, traceNonUniformIndex walks the data flow graph backwards. Originally, it just walked until the goal node. Your change in addition needs to collect all the intermediate instructions.

The waterfall loop insertion then should just operate in terms of where the cut is, and what instructions where collected in the interim (which you can think of as one side of the cut).

Really, your solution should have a single method for inserting the loop -- having separate emitWaterfallBegin and emitWaterfallBeginForScalarizedLoops methods seems wrong.

kmitropoulou commented:
Thank you for the feedback. I will see if I can make the algorithm more generic. I am looking in the failures.

kmitropoulou commented:
>>In order to do that, traceNonUniformIndex walks the data flow graph backwards. Originally, it just walked until the goal >>node. Your change in addition needs to collect all the intermediate instructions.

@nhaehnle : I push in the waterfall loop the same instructions as the original code does (before my changes). Why should we change this? How can we benefit from this?

kmitropoulou commented:
This is WIP since I want to do a bit more testing.

amdvlk-admin commented:
ba71ab4b5fa2820dc2df994e997ae480f8b58ef2 Jenkins build error.
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/disassembler/Disassembler.cpp:141:12: error: ‘endianness’ in namespace ‘llvm::support’ does not name a type; did you mean ‘endian’?
  141 |   support::endianness endian() { return m_objFile->isLittleEndian() ? support::little : support::big; }
      |            ^~~~~~~~~~
      |            endian
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/disassembler/Disassembler.cpp: In member function ‘void {anonymous}::ObjDisassembler::addBinaryEncodingComment(llvm::raw_ostream&, unsigned int, llvm::ArrayRef<unsigned char>)’:
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/disassembler/Disassembler.cpp:677:9: error: ‘endian’ was not declared in this scope
  677 |     if (endian() == support::little)
      |         ^~~~~~
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/disassembler/Disassembler.cpp:677:9: note: suggested alternatives:
In file included from /usr/include/c++/9/memory:93,
                 from /jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llvm-project/llvm/include/llvm/ADT/SmallVector.h:28,
--
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/disassembler/Disassembler.cpp:677:30: error: ‘little’ is not a member of ‘llvm::support’; did you mean ‘little_t’?
  677 |     if (endian() == support::little)
      |                              ^~~~~~
      |                              little_t
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/disassembler/Disassembler.cpp: In member function ‘size_t {anonymous}::ObjDisassembler::decodeNote(llvm::StringRef)’:
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/disassembler/Disassembler.cpp:794:60: error: ‘endian’ was not declared in this scope
  794 |   unsigned nameSize = support::endian::read32(data.data(), endian());
      |                                                            ^~~~~~
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/disassembler/Disassembler.cpp:794:60: note: suggested alternatives:
In file included from /usr/include/c++/9/memory:93,
                 from /jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llvm-project/llvm/include/llvm/ADT/SmallVector.h:28,

tsymalla commented on lgc/builder/BuilderImpl.cpp:
```diff
@@ -334,6 +335,67 @@ BranchInst *BuilderImpl::createIf(Value *condition, bool wantElse, const Twine &
+// A simple memory efficient container that holds up to 64 instructions in a bit vector. It needs two helper data
+// structures: 1. indexInsnMap that maps an instruction to its index in the bit vector and 2. insnChain that maps an
+// index back to an instruction.
+class TinyInsnSet {
```
Can we name this `TinyInstructionSet`?

kmitropoulou replied:
Done

tsymalla commented on lgc/builder/BuilderImpl.cpp:
```diff
@@ -334,6 +335,67 @@ BranchInst *BuilderImpl::createIf(Value *condition, bool wantElse, const Twine &
+class TinyInsnSet {
+  BitVector bits;
+  const DenseMap<Instruction *, unsigned> &indexInsnMap;
+  const SmallVector<Instruction *, 64> &insnChain;
```
Can you please alias that, so we don't need to have the full type signature everywhere?
using InstructionChainVec = SmallVector<Instruction *, 64>;

kmitropoulou replied:
Done

tsymalla commented on lgc/builder/BuilderImpl.cpp:
```diff
@@ -334,6 +335,67 @@ BranchInst *BuilderImpl::createIf(Value *condition, bool wantElse, const Twine &
+// index back to an instruction.
+class TinyInsnSet {
+  BitVector bits;
+  const DenseMap<Instruction *, unsigned> &indexInsnMap;
```
Maybe name this instructionIndexMap, since it maps an instruction to an index

tsymalla commented on lgc/builder/BuilderImpl.cpp:
```diff
@@ -334,6 +335,67 @@ BranchInst *BuilderImpl::createIf(Value *condition, bool wantElse, const Twine &
+// structures: 1. indexInsnMap that maps an instruction to its index in the bit vector and 2. insnChain that maps an
+// index back to an instruction.
+class TinyInsnSet {
+  BitVector bits;
```
```suggestion
  BitVector occupiedInstructionSlots;
```

kmitropoulou replied:
I prefer to use bits as bit vector does.

tsymalla commented on lgc/builder/BuilderImpl.cpp:
```diff
@@ -334,6 +335,67 @@ BranchInst *BuilderImpl::createIf(Value *condition, bool wantElse, const Twine &
+
+  class iterator {
+    BitVector::const_set_bits_iterator it;
+    const SmallVector<Instruction *, 64> &insnChain;
```
Can the iterator directly access the parent's class scope, since a nested class should be befriended to its parent class?

kmitropoulou replied:
I do not think so.

tsymalla commented on lgc/builder/BuilderImpl.cpp:
```diff
@@ -334,6 +335,67 @@ BranchInst *BuilderImpl::createIf(Value *condition, bool wantElse, const Twine &
+      ++it;
+      return *this;
+    }
+    Instruction *operator*() {
```
Can you please separate the member functions with empty lines?

kmitropoulou replied:
Done

tsymalla commented on lgc/builder/BuilderImpl.cpp:
```diff
@@ -334,6 +335,67 @@ BranchInst *BuilderImpl::createIf(Value *condition, bool wantElse, const Twine &
+    bits.set(index);
+  }
+
+  unsigned size() { return bits.size(); }
```
Can be const

kmitropoulou replied:
Done

tsymalla commented on lgc/builder/BuilderImpl.cpp:
```diff
@@ -334,6 +335,67 @@ BranchInst *BuilderImpl::createIf(Value *condition, bool wantElse, const Twine &
+
+  unsigned size() { return bits.size(); }
+
+  bool empty() { return bits.size() == 0; }
```
Can be const, plus use bits.empty()

kmitropoulou replied:
Done

amdvlk-admin commented:
a5df20e570cafbe91be3d4df84764c04229728ab Jenkins build error.
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/disassembler/Disassembler.cpp:141:12: error: ‘endianness’ in namespace ‘llvm::support’ does not name a type; did you mean ‘endian’?
  141 |   support::endianness endian() { return m_objFile->isLittleEndian() ? support::little : support::big; }
      |            ^~~~~~~~~~
      |            endian
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/disassembler/Disassembler.cpp: In member function ‘void {anonymous}::ObjDisassembler::addBinaryEncodingComment(llvm::raw_ostream&, unsigned int, llvm::ArrayRef<unsigned char>)’:
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/disassembler/Disassembler.cpp:677:9: error: ‘endian’ was not declared in this scope
  677 |     if (endian() == support::little)
      |         ^~~~~~
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/disassembler/Disassembler.cpp:677:9: note: suggested alternatives:
In file included from /usr/include/c++/9/memory:93,
                 from /jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llvm-project/llvm/include/llvm/ADT/SmallVector.h:28,
--
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/disassembler/Disassembler.cpp:677:30: error: ‘little’ is not a member of ‘llvm::support’; did you mean ‘little_t’?
  677 |     if (endian() == support::little)
      |                              ^~~~~~
      |                              little_t
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/disassembler/Disassembler.cpp: In member function ‘size_t {anonymous}::ObjDisassembler::decodeNote(llvm::StringRef)’:
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/disassembler/Disassembler.cpp:794:60: error: ‘endian’ was not declared in this scope
  794 |   unsigned nameSize = support::endian::read32(data.data(), endian());
      |                                                            ^~~~~~
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/disassembler/Disassembler.cpp:794:60: note: suggested alternatives:
In file included from /usr/include/c++/9/memory:93,
                 from /jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llvm-project/llvm/include/llvm/ADT/SmallVector.h:28,

tsymalla commented on lgc/builder/BuilderImpl.cpp:
```diff
@@ -334,6 +335,75 @@ BranchInst *BuilderImpl::createIf(Value *condition, bool wantElse, const Twine &
+
+  void insert(Instruction *instr) {
+    auto it = instrToIndex.find(instr);
+    assert(it != instrToIndex.end() && "Expected to find I in instrToIndex.");
```
```suggestion
    assert(it != instrToIndex.end() && "Expected to find instr in instrToIndex.");
```

kmitropoulou replied:
Done :)

tsymalla commented:
Thanks for addressing the comments. There is some build failure, maybe you need to rebase?

kmitropoulou commented:
> Thanks for addressing the comments. There is some build failure, maybe you need to rebase?

For two days, I cannot upload the patch for review. I had to use an older workspace.

amdvlk-admin commented:
ebe1ba5104e120e9cd1f34f73d6d5f8ae15645f4 Jenkins build error.
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/disassembler/Disassembler.cpp:141:12: error: ‘endianness’ in namespace ‘llvm::support’ does not name a type; did you mean ‘endian’?
  141 |   support::endianness endian() { return m_objFile->isLittleEndian() ? support::little : support::big; }
      |            ^~~~~~~~~~
      |            endian
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/disassembler/Disassembler.cpp: In member function ‘void {anonymous}::ObjDisassembler::addBinaryEncodingComment(llvm::raw_ostream&, unsigned int, llvm::ArrayRef<unsigned char>)’:
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/disassembler/Disassembler.cpp:677:9: error: ‘endian’ was not declared in this scope
  677 |     if (endian() == support::little)
      |         ^~~~~~
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/disassembler/Disassembler.cpp:677:9: note: suggested alternatives:
In file included from /usr/include/c++/9/memory:93,
                 from /jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llvm-project/llvm/include/llvm/ADT/SmallVector.h:28,
--
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/disassembler/Disassembler.cpp:677:30: error: ‘little’ is not a member of ‘llvm::support’; did you mean ‘little_t’?
  677 |     if (endian() == support::little)
      |                              ^~~~~~
      |                              little_t
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/disassembler/Disassembler.cpp: In member function ‘size_t {anonymous}::ObjDisassembler::decodeNote(llvm::StringRef)’:
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/disassembler/Disassembler.cpp:794:60: error: ‘endian’ was not declared in this scope
  794 |   unsigned nameSize = support::endian::read32(data.data(), endian());
      |                                                            ^~~~~~
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/disassembler/Disassembler.cpp:794:60: note: suggested alternatives:
In file included from /usr/include/c++/9/memory:93,
                 from /jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llvm-project/llvm/include/llvm/ADT/SmallVector.h:28,

amdvlk-admin commented:
### Test summary for commit 1b229475f1e5b69489439c828cac29c39d020f8e
<details><summary>CTS tests (Failed: 0/138362)</summary><ul><li>Built with version 1.3.5.2</li>
<details open><summary>Ubuntu navi3x, Srdcvk</summary><ul>

<li>Passed: 35163/69147 (50.9%)</li>
<li>Failed: 0/69147 (0.0%)</li>
<li>Not Supported: 33984/69147 (49.1%)</li><li>Warnings: 0/69147 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu navi2x, Srdcvk</summary><ul>

<li>Passed: 35242/69215 (50.9%)</li>
<li>Failed: 0/69215 (0.0%)</li>
<li>Not Supported: 33973/69215 (49.1%)</li><li>Warnings: 0/69215 (0.0%)</li>
</ul></details>


</ul></details>

amdvlk-admin commented:
### Test summary for commit 8edc895f5c3351ba125a50f9fc98d43234c2f83d
<details><summary>CTS tests (Failed: 0/138362)</summary><ul><li>Built with version 1.3.5.2</li>
<details open><summary>Ubuntu navi3x, Srdcvk</summary><ul>

<li>Passed: 35163/69147 (50.9%)</li>
<li>Failed: 0/69147 (0.0%)</li>
<li>Not Supported: 33984/69147 (49.1%)</li><li>Warnings: 0/69147 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu navi2x, Srdcvk</summary><ul>

<li>Passed: 35242/69215 (50.9%)</li>
<li>Failed: 0/69215 (0.0%)</li>
<li>Not Supported: 33973/69215 (49.1%)</li><li>Warnings: 0/69215 (0.0%)</li>
</ul></details>


</ul></details>

amdvlk-admin commented:
### Test summary for commit 5f4d45a748fef604d7a8990b488210271abb65f8
<details><summary>CTS tests (Failed: 0/138362)</summary><ul><li>Built with version 1.3.5.2</li>
<details open><summary>Ubuntu navi3x, Srdcvk</summary><ul>

<li>Passed: 35162/69147 (50.9%)</li>
<li>Failed: 0/69147 (0.0%)</li>
<li>Not Supported: 33985/69147 (49.1%)</li><li>Warnings: 0/69147 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu navi2x, Srdcvk</summary><ul>

<li>Passed: 35242/69215 (50.9%)</li>
<li>Failed: 0/69215 (0.0%)</li>
<li>Not Supported: 33973/69215 (49.1%)</li><li>Warnings: 0/69215 (0.0%)</li>
</ul></details>


</ul></details>

nhaehnle commented on llpc/context/llpcPipelineContext.cpp:
```diff
@@ -610,13 +610,12 @@ ShaderOptions PipelineContext::computeShaderOptions(const PipelineShaderInfo &sh
-    // Enable waterfall load scalarization when vgpr limit is set.
-    if (shaderOptions.vgprLimit != 0 && shaderOptions.vgprLimit != UINT_MAX)
-      shaderOptions.scalarizeWaterfallLoads = true;
+  else {
```
We should usually put { } around both sides of an if/else if one of them needs it.

kmitropoulou replied:
Done

nhaehnle reviewed (commented):
Thanks for working towards the next round on this! :)

Couple of high-level points that I think must be addressed before more details:

* The interface of `traceNonUniformIndex` is too complex. The interface prior to this patch is: non-uniform value in, earlier non-uniform value (aka "index") out. The new interface should just be: non-uniform value in, earlier non-uniform value (aka "index") plus list of instruction to be cloned out.
* Do you really need those references in TinyInstructionSet? They're unidiomatic and wasteful.
* There are all these calls to `insertNewValueInInsnDeps`. It should be possible to do all that from `propagate`.

kmitropoulou commented:
>> Do you really need those references in TinyInstructionSet? They're unidiomatic and wasteful.

I thought that we wanted to use a memory efficient container. Otherwise, I can just use std::set .

kmitropoulou commented:
> > > Do you really need those references in TinyInstructionSet? They're unidiomatic and wasteful.
> 
> I thought that we wanted to use a memory efficient container. Otherwise, I can just use std::set .

I am sorry I am confused. What do you mean?

kmitropoulou commented:
> > > > Do you really need those references in TinyInstructionSet? They're unidiomatic and wasteful.
> > 
> > 
> > I thought that we wanted to use a memory efficient container. Otherwise, I can just use std::set .
> 
> I am sorry I am confused. What do you mean?

I will try to hide TinyInstructionSet::InstrToIndexMap &instrToIndex and TinyInstructionSet::IndexToInstructionVec &indexToInstr.

amdvlk-admin commented:
### Test summary for commit a47200fbf7659dc2bae6328789908b72122f43a8
<details><summary>CTS tests (Failed: 0/138443)</summary><ul><li>Built with version 1.3.5.2</li>
<details open><summary>Ubuntu navi3x, Srdcvk</summary><ul>

<li>Passed: 35211/69228 (50.9%)</li>
<li>Failed: 0/69228 (0.0%)</li>
<li>Not Supported: 34017/69228 (49.1%)</li><li>Warnings: 0/69228 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu navi2x, Srdcvk</summary><ul>

<li>Passed: 35242/69215 (50.9%)</li>
<li>Failed: 0/69215 (0.0%)</li>
<li>Not Supported: 33973/69215 (49.1%)</li><li>Warnings: 0/69215 (0.0%)</li>
</ul></details>


</ul></details>

amdvlk-admin commented:
### Test summary for commit 43021f8e9632009f087bf6bcc2b28f92800784a4
<details><summary>CTS tests (Failed: 0/138443)</summary><ul><li>Built with version 1.3.5.2</li>
<details open><summary>Ubuntu navi3x, Srdcvk</summary><ul>

<li>Passed: 35211/69228 (50.9%)</li>
<li>Failed: 0/69228 (0.0%)</li>
<li>Not Supported: 34017/69228 (49.1%)</li><li>Warnings: 0/69228 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu navi2x, Srdcvk</summary><ul>

<li>Passed: 35242/69215 (50.9%)</li>
<li>Failed: 0/69215 (0.0%)</li>
<li>Not Supported: 33973/69215 (49.1%)</li><li>Warnings: 0/69215 (0.0%)</li>
</ul></details>


</ul></details>

kmitropoulou commented:
> > > > > Do you really need those references in TinyInstructionSet? They're unidiomatic and wasteful.
> > > 
> > > 
> > > I thought that we wanted to use a memory efficient container. Otherwise, I can just use std::set .
> > 
> > 
> > I am sorry I am confused. What do you mean?
> 
> I will try to hide TinyInstructionSet::InstrToIndexMap &instrToIndex and TinyInstructionSet::IndexToInstructionVec &indexToInstr.

Do you mean the references const InstrToIndexMap &instrToIndex and const IndexToInstructionVec &indexToInstr? Removing them from the class will save space, but I will have to pass them as arguments in insert(). What do you think?

amdvlk-admin commented:
### Test summary for commit 36105412d756be256595545f2d8f19663a103c01
<details><summary>CTS tests (Failed: 0/138443)</summary><ul><li>Built with version 1.3.5.2</li>
<details open><summary>Ubuntu navi3x, Srdcvk</summary><ul>

<li>Passed: 35211/69228 (50.9%)</li>
<li>Failed: 0/69228 (0.0%)</li>
<li>Not Supported: 34017/69228 (49.1%)</li><li>Warnings: 0/69228 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu navi2x, Srdcvk</summary><ul>

<li>Passed: 35242/69215 (50.9%)</li>
<li>Failed: 0/69215 (0.0%)</li>
<li>Not Supported: 33973/69215 (49.1%)</li><li>Warnings: 0/69215 (0.0%)</li>
</ul></details>


</ul></details>

tsymalla commented on llpc/test/shaderdb/core/OpTypeSampledImage_TestWaterfallScalarize.frag:
```diff
@@ -16,18 +13,56 @@ void main()
-// BEGIN_SHADERTEST
-//
-// RUN: amdllpc -scalarize-waterfall-descriptor-loads -v %gfxip %s | FileCheck -check-prefix=SHADERTEST %s
+// RUN: amdllpc -scalarize-waterfall-descriptor-loads -v %gfxip %s | FileCheck -check-prefix=GFX %s
```
Why the different test prefix by the way?

kmitropoulou replied:
Done :) Thank you! I did not know that we have a standard name for it.

kmitropoulou commented:
https://github.com/GPUOpen-Drivers/llpc/pull/2759#discussion_r1391667025

Removed it

kmitropoulou commented:
> Thanks for working towards the next round on this! :)
> 
> Couple of high-level points that I think must be addressed before more details:
> 
> * The interface of `traceNonUniformIndex` is too complex. The interface prior to this patch is: non-uniform value in, earlier non-uniform value (aka "index") out. The new interface should just be: non-uniform value in, earlier non-uniform value (aka "index") plus list of instruction to be cloned out.
> * Do you really need those references in TinyInstructionSet? They're unidiomatic and wasteful.
> * There are all these calls to `insertNewValueInInsnDeps`. It should be possible to do all that from `propagate`.

You are right it was not an overkill to have 2 64-bit references in TinyInstructionSet. I removed them.

kmitropoulou commented:
I am still working on scalarizationOfDescriptorLoadsNegativeTest1.ll and scalarizationOfDescriptorLoadsNegativeTest2.ll

amdvlk-admin commented:
### Test summary for commit 43a7a13adabee576ec46f9534b55c986402b3b4a
<details><summary>CTS tests (Failed: 0/138443)</summary><ul><li>Built with version 1.3.5.2</li>
<details open><summary>Ubuntu navi3x, Srdcvk</summary><ul>

<li>Passed: 35211/69228 (50.9%)</li>
<li>Failed: 0/69228 (0.0%)</li>
<li>Not Supported: 34017/69228 (49.1%)</li><li>Warnings: 0/69228 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu navi2x, Srdcvk</summary><ul>

<li>Passed: 35242/69215 (50.9%)</li>
<li>Failed: 0/69215 (0.0%)</li>
<li>Not Supported: 33973/69215 (49.1%)</li><li>Warnings: 0/69215 (0.0%)</li>
</ul></details>


</ul></details>

tsymalla commented on lgc/builder/BuilderImpl.cpp:
```diff
@@ -329,6 +330,113 @@ BranchInst *BuilderImpl::createIf(Value *condition, bool wantElse, const Twine &
+    // Initialization of instrToIndex and indexToInstr.
+    if (scalarizeDescriptorLoads) {
+      unsigned cnt = 0;
+      for (Instruction *I = nonUniformInst->getPrevNode(); I != nullptr && cnt < 64; I = I->getPrevNode(), ++cnt) {
```
Is the maximum number of steps something we should parametrize (or at least provide it as easily parametrizable value)?

kmitropoulou replied:
Done.

tsymalla commented on lgc/builder/BuilderImpl.cpp:
```diff
@@ -329,6 +330,113 @@ BranchInst *BuilderImpl::createIf(Value *condition, bool wantElse, const Twine &
+    return;
+  }
+  assert(instrDeps.contains(currentVisitedInsn) && "The current visited instruction should have been in the map.");
+  auto it1 = instrDeps.try_emplace(newValue).first;
```
Can you use a structured binding so this is a bit cleaner?

kmitropoulou replied:
Done.

tsymalla replied:
Yes, but you can check for the boolean value directly. However, it is just personal taste.

kmitropoulou replied:
I am not using the boolean. That's why I did not add it.

tsymalla commented on lgc/builder/BuilderImpl.cpp:
```diff
@@ -329,6 +330,113 @@ BranchInst *BuilderImpl::createIf(Value *condition, bool wantElse, const Twine &
+  auto &setOfInsns = it1->second;
+  auto it2 = instrDeps.find(currentVisitedInsn);
+  const auto &set = it2->second;
+  for (auto it3 = set.begin(indexToInstr), ite = set.end(indexToInstr); it3 != ite; ++it3) {
```
Can you use a range based for loop instead?

kmitropoulou replied:
I did not do it because begin() and end() of TinyInstructionSet have an argument. Please let me know if there is a way to do it.

tsymalla replied:
I see, you originally stored indexToInstr as a member reference, right? With that approach, it would have worked

kmitropoulou replied:
Exactly :)

tsymalla replied:
How about not storing the reference to indexToInstr and instead add a separate member function that returns an `iterator_range` on which you can iterate?

````
iterator_range<IteratorT> getIteratorRange(IndexToInstrMap &indexToInstr) {
return iterator_range(set.begin(indexToInstr), set.end(indexToInstr));
}
```

That should allow you to make use of the range based for loop and not increase the storage size of the TinyInstructionSet?

amdvlk-admin commented:
### Test summary for commit bbe0a573c41f0ecef4b3df729adc48bdfdf8b705
<details><summary>CTS tests (Failed: 0/138443)</summary><ul><li>Built with version 1.3.5.2</li>
<details open><summary>Ubuntu navi3x, Srdcvk</summary><ul>

<li>Passed: 35211/69228 (50.9%)</li>
<li>Failed: 0/69228 (0.0%)</li>
<li>Not Supported: 34017/69228 (49.1%)</li><li>Warnings: 0/69228 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu navi2x, Srdcvk</summary><ul>

<li>Passed: 35242/69215 (50.9%)</li>
<li>Failed: 0/69215 (0.0%)</li>
<li>Not Supported: 33973/69215 (49.1%)</li><li>Warnings: 0/69215 (0.0%)</li>
</ul></details>


</ul></details>

amdvlk-admin commented:
### Test summary for commit 8179bc4cb333f96f41af18ad8c2d8eda9ea895d7
<details><summary>CTS tests (Failed: 0/138378)</summary><ul><li>Built with version 1.3.5.2</li>
<details open><summary>Ubuntu navi3x, Srdcvk</summary><ul>

<li>Passed: 35162/69163 (50.8%)</li>
<li>Failed: 0/69163 (0.0%)</li>
<li>Not Supported: 34001/69163 (49.2%)</li><li>Warnings: 0/69163 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu navi2x, Srdcvk</summary><ul>

<li>Passed: 35242/69215 (50.9%)</li>
<li>Failed: 0/69215 (0.0%)</li>
<li>Not Supported: 33973/69215 (49.1%)</li><li>Warnings: 0/69215 (0.0%)</li>
</ul></details>


</ul></details>

amdvlk-admin commented:
### Test summary for commit 504d64b39c1cb5f64c5866e17b77521e51b82f3f
<details><summary>CTS tests (Failed: 0/138443)</summary><ul><li>Built with version 1.3.5.2</li>
<details open><summary>Ubuntu navi3x, Srdcvk</summary><ul>

<li>Passed: 35211/69228 (50.9%)</li>
<li>Failed: 0/69228 (0.0%)</li>
<li>Not Supported: 34017/69228 (49.1%)</li><li>Warnings: 0/69228 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu navi2x, Srdcvk</summary><ul>

<li>Passed: 35241/69215 (50.9%)</li>
<li>Failed: 0/69215 (0.0%)</li>
<li>Not Supported: 33974/69215 (49.1%)</li><li>Warnings: 0/69215 (0.0%)</li>
</ul></details>


</ul></details>

piotrAMD commented on lgc/builder/BuilderImpl.cpp:
```diff
@@ -329,6 +330,114 @@ BranchInst *BuilderImpl::createIf(Value *condition, bool wantElse, const Twine &
+void TraceNonUniformIndex::insertNewValueInInstrDeps(Value *newValue, Instruction *currentVisitedInstr) {
+  if (!instrToIndex.contains(currentVisitedInstr)) {
+    // The instruction is either outside of 64 limit or in a different basic block. So, we bail-out scalarization.
+    scalarizeDescriptorLoads = false;
```
It seems setting TraceNonUniformIndex.scalarizeDescriptorLoads to `false` here means that the next time TraceNonUniformIndex::run is invoked it will not try to scalarize. That means if there are two non-uniform operands of, say, image.sample, if the first one could not be scalarized, then the second one would not be either.

kmitropoulou replied:
This is how the existing code works. I am working to change it.

piotrAMD replied:
Fair enough. Can you extend the comment to make it clear that this will also affect scalarization of other operands?

Implementation-wise I guess it is fine to address it in a follow-up commit to keep the scope of the changes smaller.

kmitropoulou replied:
Done in https://github.com/GPUOpen-Drivers/llpc/pull/2859

piotrAMD commented on lgc/builder/BuilderImpl.cpp:
```diff
@@ -329,6 +330,114 @@ BranchInst *BuilderImpl::createIf(Value *condition, bool wantElse, const Twine &
+
+void TraceNonUniformIndex::insertNewValueInInstrDeps(Value *newValue, Instruction *currentVisitedInstr) {
+  if (!instrToIndex.contains(currentVisitedInstr)) {
+    // The instruction is either outside of 64 limit or in a different basic block. So, we bail-out scalarization.
```
> or in a different basic block

I know this is a part of the design of this patch, but do we really have to give up the scalarization if an instruction is in a different block?

A descriptor load in a different basic block than the image op is quite common, and with the changes here we would stop scalarizing it.

kmitropoulou replied:
It is not part of the design of this patch. This is how the existing code (before my changes) work because there is not backend support for it.

piotrAMD replied:
Yes, sorry, I misspoke - this is not really related to your patch.

kmitropoulou replied:
Done in https://github.com/GPUOpen-Drivers/llpc/pull/2877

nhaehnle commented on lgc/include/lgc/builder/BuilderImpl.h:
```diff
@@ -58,6 +58,9 @@ class BuilderImpl : public BuilderDefs {
                                          const llvm::Twine &instName = "");
 
+  // llvm::Value *createWaterfallLastUse(llvm::Value *waterfallBegin, llvm::Value *nonUniformOperand,
+  //                                     bool useVgprForOperands, const llvm::Twine &instName = "");
```
Please remove this entirely. We shouldn't leave commented-out code around.

nhaehnle reviewed (commented):
I think the code changes are good. I think Piotr's points are fair but let's address them in separate PRs that you can work on while we're working out the last remaining kinks here.

There's still a thing I noticed inline, and the LGC tests ought to be simplified. As-is, they're touching a bunch of unrelated things like reading VS inputs. Please take a look at other tests, especially in lgc/test/Transforms/{Combine,Lower}CooperativeMatrix for examples of tests that are focused on shorter instruction sequences.

kmitropoulou commented:
@nhaehnle : I am working on refactoring the tests.

kmitropoulou commented:
> @nhaehnle : I am working on refactoring the tests.

Done :)

amdvlk-admin commented:
### Test summary for commit 195b936b8b8deb5b68a3e7d19d777dea8ea85415
<details><summary>CTS tests (Failed: 0/138443)</summary><ul><li>Built with version 1.3.5.2</li>
<details open><summary>Ubuntu navi3x, Srdcvk</summary><ul>

<li>Passed: 35211/69228 (50.9%)</li>
<li>Failed: 0/69228 (0.0%)</li>
<li>Not Supported: 34017/69228 (49.1%)</li><li>Warnings: 0/69228 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu navi2x, Srdcvk</summary><ul>

<li>Passed: 35242/69215 (50.9%)</li>
<li>Failed: 0/69215 (0.0%)</li>
<li>Not Supported: 33973/69215 (49.1%)</li><li>Warnings: 0/69215 (0.0%)</li>
</ul></details>


</ul></details>

kmitropoulou commented:
I refactored the tests as much as I could.

kmitropoulou commented:
ping
